### PR TITLE
JDK-8276422 Add command-line option to disable finalization

### PR DESCRIPTION
**Pull request by stuart-marks**
Pretty much what it says. The new option controls a static member in InstanceKlass that's consulted to determine whether the finalization machinery is activated for instances when a class is loaded. A new native method is added so that this state can be queried from Java. This is used to control whether a finalizer thread is created and to disable the `System` and `Runtime::runFinalization` methods. Includes tests for the above.

Adding an option to disable finalization is part of [JEP 421](https://openjdk.java.net/jeps/421).

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8276422](https://bugs.openjdk.java.net/browse/JDK-8276422): Add command-line option to disable finalization

### Reviewers
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Kim Barrett](https://openjdk.java.net/census#kbarrett) (@kimbarrett - **Reviewer**)
 * [Brent Christian](https://openjdk.java.net/census#bchristi) (@bchristi-git - **Reviewer**)

### Contributors
 * David Holmes `<dholmes@openjdk.org>`
 * Brent Christian `<bchristi@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6442/head:pull/6442` \
`$ git checkout pull/6442`

Update a local copy of the PR: \
`$ git checkout pull/6442` \
`$ git pull https://git.openjdk.java.net/jdk pull/6442/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6442`

View PR using the GUI difftool: \
`$ git pr show -t 6442`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6442.diff">https://git.openjdk.java.net/jdk/pull/6442.diff</a>

</details>

### CODE REVIEW

**Review by dholmes-ora — APPROVED** [review]
Hi Stuart,

This all looks fine to me. The hotspot part needs a second reviewer (especially as I contributed a chunk of that code :) ).

Thanks,
David

File: src/java.base/share/classes/java/lang/ref/Finalizer.java
Comment (jaikiran) [inline src/java.base/share/classes/java/lang/ref/Finalizer.java:195]: Hello Stuart,
My understanding of the the lazy `Holder` is that it's there to delay the static initialization of whatever is part of that static block. In this case here, the `Holder` is being used within the `static` block of the `Finalizer` class, that too as the first thing. In this case, is that `Holder` class necessary?

Comment (stuart-marks) [reply src/java.base/share/classes/java/lang/ref/Finalizer.java:195]: Huh, good catch! This was mostly left over from an earlier version of the flag that used system properties, which aren't initialized until after the Finalizer class is initialized.

It might be the case that the Holder can be removed at this point, since the finalization-enabled bit is no longer in a system property and is in a native class member that should be available before the VM is started.

I say "might" though because this occurs early in system startup, and weird things potentially happen. For example, suppose the first object with a finalizer is created before the Finalizer class is initialized. The VM will perform an upcall to Finalizer::register. An ordinary call to a static method will ensure the class is initialized before proceeding with the call, but this VM upcall is a special case.... I'll have to investigate this some more.

Comment (dholmes-ora) [reply src/java.base/share/classes/java/lang/ref/Finalizer.java:195]: @stuart-marks not sure I see how anything is different here compared to the existing logic. The `Finalizer` class is explicitly initialized quite early in the init process, but if a preceding class's initialization created an object with a finalizer then that same upcall would be involved.

**Review by kimbarrett — CHANGES_REQUESTED** [review]
I only really reviewed the hotspot changes.

There is nothing here to make the various GCs take advantage of finalization being disabled.  Is the plan to leave that to followup changes?

File: src/java.base/share/classes/java/lang/ref/Finalizer.java
Comment (shipilev) [reply src/java.base/share/classes/java/lang/ref/Finalizer.java:195]: Do we even have to have a flag on Java side? It looks like these calls are only done as the upcalls from VM, so we might just keep the flag on VM side?

**Review by shipilev — COMMENTED** [review]
From the brief look, it is OK. Minor nits.

File: src/java.base/share/classes/java/lang/ref/Finalizer.java
Comment (dholmes-ora) [reply src/java.base/share/classes/java/lang/ref/Finalizer.java:195]: @shipilev not sure what you mean by  "a flag on the Java side". The Java code just queries the VM for the finalization enabled/disabled state and uses that to control things.

Comment (shipilev) [reply src/java.base/share/classes/java/lang/ref/Finalizer.java:195]: Yeah, "flag" is `Holder.ENABLED` here. I mean, are Java methods `registerFinalizer` and `runFinalization` called only by VM? If so, can VM check the whole thing on VM side, without going to Java and asking back from there?

File: src/hotspot/share/oops/instanceKlass.hpp
Comment (dholmes-ora) [reply src/hotspot/share/oops/instanceKlass.hpp:338]: This was intended as an accessor function, similar to `count()` or `offset()` not a query as-in `is_shared_boot_class()`. As it is a boolean field you could convert it to a query instead.

File: src/java.base/share/classes/java/lang/ref/Finalizer.java
Comment (dholmes-ora) [reply src/java.base/share/classes/java/lang/ref/Finalizer.java:195]: `registerFinalizer` does not expect to be called and only uses the "flag" as a form of assertion.

`runFinalization` is called from Java code.

**Comment by dholmes-ora** [comment]
> There is nothing here to make the various GCs take advantage of finalization being disabled. Is the plan to leave that to followup changes?

@kimbarrett I provided the basic VM parts here. I'm not aware of what specifically a GC might optimise if it knows there can be no finalizers, but that seems like something the GC folk should look to providing as a follow up. Thanks.

File: src/java.base/share/classes/java/lang/ref/Finalizer.java
Comment (AlanBateman) [reply src/java.base/share/classes/java/lang/ref/Finalizer.java:195]: I think @shipilev asks a good question. This could be done completely in the VM without the changes to j.l.ref.Finalizer. The CLI option is for experimenting, at least in the short term, and should be benign to have the Finalizer thread running, it just won't do anything.

Comment (plevart) [reply src/java.base/share/classes/java/lang/ref/Finalizer.java:195]: Or, you could move the static initialization block that statrts the finalizer thread into the Finalizer.FinalizerThread class itself and then arrange for that class to be initialized explicitly immediately after the Finalizer class, but conditionally, only if the option to disable finalization was not specified...
This way the Finalizer class could still be initialized early, but the thread would not be started if it is not needed.

Comment (stuart-marks) [reply src/java.base/share/classes/java/lang/ref/Finalizer.java:195]: The disable-finalization feature is a bit more than experimental. The goal is to provide a faithful representation of what the system will look like when finalization is removed. Of course most of that is objects' `finalize` methods not being called, but it also includes having no finalizer thread running, as well as having `runFinalization` (a public API) do nothing at all.

Comment (stuart-marks) [reply src/java.base/share/classes/java/lang/ref/Finalizer.java:195]: @dholmes-ora If the Finalizer class is initialized explicitly and at the right time, then maybe we can do away with the Holder class entirely. Can you point me to where this is done?

File: src/hotspot/share/prims/jvm.cpp
Comment (mlchung) [reply src/hotspot/share/prims/jvm.cpp]: One typical way for VM to pass the arguments to the library is via private system properties.   System::initPhase1 will save the VM properties in `jdk.internal.misc.VM` and filters out the private properties from the system properties returned from System::getProperties (see System::createProperties).

You can query the flag via `jdk.internal.misc.VM.getProperty("jdk.finalization.disabled")` for example.

I don't see any issue moving the Finalizer class initialization after initPhase1 since there is no finalizer during VM startup.

**Comment by mlchung** [comment]
When the finalization is disabled, perhaps jcmd GC.finalizer_info should just be made as a nop in the VM.

File: src/hotspot/share/prims/jvm.cpp
Comment (stuart-marks) [reply src/hotspot/share/prims/jvm.cpp]: I renamed the function to `is_finalization_enabled` per previous comment, and I also made these cleanups.

Comment (stuart-marks) [reply src/hotspot/share/prims/jvm.cpp]: Fixed.

Comment (stuart-marks) [reply src/hotspot/share/prims/jvm.cpp]: Regarding using system properties, my initial prototype did this in the launcher, and it did run into the problem that the Finalizer class is initialized before system properties are available. That's why I created the Holder class, so that reading the property could be delayed until the first upcall to Finalizer::register. I suppose the initialization of Finalizer could be moved later, but that seems more invasive.

The flag needs to be available in the VM in order to avoid upcalls for instances-with-finalizers in the first place. Alan had [suggested](https://bugs.openjdk.java.net/browse/JDK-8276422?focusedCommentId=14456185&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-14456185) moving the argument processing into the VM, and David suggested putting the flag into InstanceKlass, which seems a sensible place to me. It's also reasonably accessible there to GC implementations, should they want to inspect it.

Comment (mlchung) [reply src/hotspot/share/prims/jvm.cpp]: > Alan had suggested moving the argument processing into the VM, and David suggested putting the flag into InstanceKlass, which seems a sensible place to me. It's also reasonably accessible there to GC implementations, should they want to inspect it.

That's still all good.  What I meant is for the VM to add a private system property (not the launcher) as to pass the flag to the library code.  The precedence is like `sun.nio.MaxDirectMemorySize`  or `java.lang.Integer.IntegerCache.high`.

**Review by dholmes-ora — APPROVED** [review]

**Comment by bchristi-git** [comment]
> When the finalization is disabled, perhaps jcmd GC.finalizer_info should just be made as a nop in the VM.

Would it be interesting (perhaps in a follow-up) for GC.finalizer_info to report that the given VM had finalization disabled?

File: src/java.base/share/classes/java/lang/ref/Finalizer.java
Comment (stuart-marks) [reply src/java.base/share/classes/java/lang/ref/Finalizer.java]: I pushed an update to remove the Holder class. It seems to continue to work fine. Thanks for pointing this out @jaikiran !

File: src/java.base/share/classes/java/lang/ref/Finalizer.java
Comment (dholmes-ora) [inline src/java.base/share/classes/java/lang/ref/Finalizer.java:64]: private?

**Review by dholmes-ora — APPROVED** [review]
Good simplification.

**Comment by dholmes-ora** [comment]
> When the finalization is disabled, perhaps jcmd GC.finalizer_info should just be made as a nop in the VM.

Yes that is a trivial change to add. @stuart-marks I can provide the code. You can choose whether to include in this PR or else we can do a follow-up.

**Comment by dholmes-ora** [comment]
@stuart-marks : https://github.com/openjdk/jdk/pull/6469  (didn't intend to actually make a PR but clicked the wrong part of the button :) )

**Comment by stuart-marks** [comment]
> > When the finalization is disabled, perhaps jcmd GC.finalizer_info should just be made as a nop in the VM.
> 
> Yes that is a trivial change to add. @stuart-marks I can provide the code. You can choose whether to include in this PR or else we can do a follow-up.

Seems simple enough. Is there any testing that needs to be done for this? Does jcmd output require CSR review? I guess there would be a compatibility issue if there were something that was parsing the output of jcmd. Or is it solely intended to be read by humans?

File: src/java.base/share/classes/java/lang/ref/Finalizer.java
Comment (stuart-marks) [reply src/java.base/share/classes/java/lang/ref/Finalizer.java:64]: Yeah, probably should be private. Other stuff in this class is private except things that are used from outside.

File: src/java.base/share/classes/java/lang/ref/Finalizer.java
Comment (jaikiran) [reply src/java.base/share/classes/java/lang/ref/Finalizer.java]: Thank you Stuart, this changed version looks fine to me.

**Comment by dholmes-ora** [comment]
@stuart-marks No CSR needed for this as no output format is specified. Plus this command already has a simple text response when there are no finalizers queued. E.g.
``` 
> ../build/linux-x64-debug-finalization/images/jdk/bin/jcmd 27939 GC.finalizer_info
27939:
No instances waiting for finalization found
```
so when finalization is disabled this just becomes:
``` 
> ../build/linux-x64-debug-finalization/images/jdk/bin/jcmd 28018 GC.finalizer_info
28018:
Finalization is disabled
```
There is a test for this Dcmd, but it doesn't test the "nothing here" case so I don't think it is necessary to augment it for this case:
`hotspot/jtreg/serviceability/dcmd/gc/FinalizerInfoTest.java`

**Review by kimbarrett — APPROVED** [review]

**Comment by stuart-marks** [comment]
Regarding **jcmd** updates, I'm thinking maybe this would be better handled separately. There is the potential to update to `GC.finalizer_info` discussed previously. Looking at the **jcmd** tool docs, it seems like `GC.run_finalization` also ought to be updated. And maybe one or more of the other commands (maybe `VM.flags` or `VM.info`?) ought to list the finalization enabled or disabled status. And of course the tool's doc will need to be updated as well.

**Review by bchristi-git — APPROVED** [review]
Lib changes and tests look good

**Comment by dholmes-ora** [comment]
@stuart-marks no issue with doing dcmd/jcmd changes separately, but I don't think we need to go too far with this. I had considered `GC.run_finalization` but it just says it calls `System.run_finalization` - so no change needed there as it will be documented in System.runFinalization. And `VM.flags` only reports `-XX` flag information. And `VM.info` doesn't seem appropriate for mentioning this either. So no further changes needed to the other Dcmds IMO and no need to update anything on the jcmd tool page either.

**Comment by stuart-marks** [comment]
@dholmes-ora OK if you're confident that it's sufficient just to add `GC.finalizer_info` and nothing else, and no docs or additional testing, then I'll just drop in the code from that branch you posted. Of course I'll do a full build & test.

**Pull request closed without merging**
